torch/distributed/device_mesh.py

```diff
@@ -217,7 +217,7 @@ def __init__(
                 "Please use a non-overlapping layout when creating a DeviceMesh."
             )
             # Because we still need to support slicing of flattened dim from root mesh, so we don't check stride here.
-            assert self._layout.numel() == self.mesh.numel(), (
+            assert self._layout.top_level_sizes == self.mesh.size(), (
```

Comment:
Once I made this check somewhat stricter, the TestDeviceMeshGetItem.test_flatten_mesh_3d testcase started failing, because of a mismatch in ndim.